### ISSUE TITLE
Glowroot Shared Lib Folder

### ISSUE THREAD

**Issue by JavaGuru-04** [glowroot/glowroot#1007]
Hello Team,
How could I use shared lib folder for  more than one microservices those are using glowroot javaagent.
I tried many ways but unable to run by pointing custom lib directory.
Currently I need to put glowroot.jar, properties file and lib folder for single microservice which in overhead if you have 10 services.

Also if same & single glowroot.jar can be used for different services ?

Objective : glowroot.properties file service specific with common glowroot.jar & lib folder.

**Comment by Nowheresly** [comment]
Hi @JavaGuru-04 
Have you tried what is explained [here](https://github.com/glowroot/glowroot/wiki/Agent-Installation-%28with-Embedded-Collector%29#running-multiple-agents-on-a-single-box)? Maybe the `multi.dir` option can help ?

**Comment by Never-lab** [comment]
Yes — you can share **one** `glowroot.jar` + `lib/` across many microservices. Each JVM needs its **own writable state** (and a unique `agent.id` if talking to Central). That matches what @nowheresly pointed at with [`multi.dir`](https://github.com/glowroot/glowroot/wiki/Agent-Installation-(with-Embedded-Collector)#running-multiple-agents-on-a-single-box).

### Pattern A — several JVMs on one host (embedded or Central)

Shared install, e.g. `/opt/glowroot/`:

```
/opt/glowroot/
  glowroot.jar
  lib/
  glowroot.properties    # multi.dir=true  (and collector.address=… if using Central)
```

`glowroot.properties` (shared):

```
multi.dir=true
# if Central:
# collector.address=http://glowroot-central:8181
```

Per service JVM:

```
-javaagent:/opt/glowroot/glowroot.jar
-Dglowroot.agent.id=orders-service
```

Glowroot then uses a **subdirectory per `agent.id`** under the install for that agent’s config/data/logs. Same jar + lib for all; no copy of `lib/` per service.

Notes:

- `agent.id` must be unique per JVM.
- Embedded UI: each agent needs its **own UI port** (set under Configuration → Web / `admin.json` in that agent’s subdir).
- The OS user of each JVM needs write access under `/opt/glowroot/` (for those subdirs / tmp).

### Pattern B — separate hosts / containers, still one shared artifact

You can still point every service at the **same read-only** `glowroot.jar` (and `lib/`) path (NFS image layer, shared volume, etc.). Put **per-service** writable dirs via system properties (or a tiny shared `glowroot.properties` + overrides):

```
-javaagent:/opt/glowroot/glowroot.jar
-Dglowroot.agent.id=orders-service
-Dglowroot.conf.dir=/var/glowroot/orders/conf
-Dglowroot.log.dir=/var/glowroot/orders/log
-Dglowroot.tmp.dir=/var/glowroot/orders/tmp
# embedded only:
-Dglowroot.data.dir=/var/glowroot/orders/data
```

For Central, `collector.address` can live in the shared properties file; override per service with `-Dglowroot.collector.address=…` if needed. There is no need for a private copy of `lib/` per microservice.

### Answers to your questions

| Question | Answer |
|----------|--------|
| Shared `lib/` + one `glowroot.jar`? | **Yes** (read-only share is fine). |
| Service-specific `glowroot.properties`? | Use **`multi.dir` + `-Dglowroot.agent.id`**, and/or `-Dglowroot.conf.dir=…` / property overrides — not a custom “lib directory” switch. |
| Same jar for different services? | **Yes.** |

### Status

Support / how-to, answered May 2023, no follow-up. Fine to close as answered.

@nowheresly — I can’t close issues; please close if you agree.
